contracts/token/ERC721/extensions/ERC721Wrapper.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "../ERC721.sol";
+import "../utils/ERC721Holder.sol";
+
+/**
+ * @dev Extension of the ERC721 token contract to support token wrapping.
+ *
+ * Users can deposit and withdraw an "underlying token" and receive a "wrapped token" with a matching tokenId. This is useful
+ * in conjunction with other modules. For example, combining this wrapping mechanism with {ERC721Votes} will allow the
+ * wrapping of an existing "basic" ERC721 into a governance token.
+ *
+ * _Available since v4.9.0_
+ */
+abstract contract ERC721Wrapper is ERC721, ERC721Holder {
+    IERC721 public immutable underlying;
+
+    constructor(IERC721 underlyingToken) {
+        underlying = underlyingToken;
+    }
+
+    /**
+     * @dev Allow a user to deposit an underlying token and mint the corresponding tokenId.
+     */
+    function depositFor(address account, uint256 tokenId) public virtual returns (bool) {
```

Comment:
Since this would only allow for a single wrapped token creation at a time, shouldn't we add a batch mechanism to allow multiple creation given an array of tokenIds?

Reply:
Same for `recover`

Reply:
If someone needs to migrate multiple tokens, it would be cheaper to do it all in one transaction even if there is no native batch transfer in ERC721 (due to warm vs. cold state). Batching could be handled either by `Multicall` or by making the argument to this function an array of `tokenIds`. I feel like the latter is simple enough that I would make that change. Do we see reasons not to?

Reply:
Make sense to me. I added `tokenIds` to both `depositFor` and `withdrawTo` as discussed.
Updated tests accordingly.